src/components/application_manager/doc/SDL.SDD.StateController.dox

```diff
@@ -0,0 +1,73 @@
+- \subpage state_controller_hight_level_logic
+
+*/
+
+/**
+\page state_controller_hight_level_logic 1 State Controller hight level logic
+
+## Introduction
+The state controller is responsible for keeping applications inappropriate HMI states 
+HMI state contains: 
+ - HMILevel 
+ - AudioStreamingState
+ - VideoStreamingState
+ - SystemContext 
+ 
+SDL should take care of conflicts between the HMI state of different apps. 
+For example, 2 apps cannot be in FULL at the same time. So if one application goes FULL, another one should be moved to BACKGROUND or LIMITED.
+Rules are described in the scope of the proposal https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0150-video-streaming-state.md 
+
+SDL should keep resolve application HMI states to avoid conflicts. There are several triggers for HMI state change:
+ - application activation
+ - application deactivation
+ - Start System Event from HMI (Phone call, TTS, etc ...)
+ - End System Event from HMI 
+ - Resumption
+
+\link application_manager::StateController 
+\endlink is responcible for this.
+
+Event from HMI should change HMI state for app registered applications.
+Changing rules described in  https://github.com/smartdevicelink/sdl_evolution/blob/proposals/0150-video-streaming-state.md 
+Some modifications introduced within https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0264-Separating-the-change-of-Audible-status-and-the-change-of-HMI-Status.md 
+
+Events from HMI are temporary and will be ended.
+After finishing of system Event all applications should receive a previous HMI state. Sp SDL should keep it.
+
+Multiple system events can present at the same time. SDL should calculate HMI state according to all of them.
+
+## Technical solution
+
+To track complicated logic of #application_manager::HmiState within system events SDL put events in internal "stack" of HMI States.
+
+There is a Regular HMI state of the application (STATE_ID_REGULAR).
+The regular state contains HMI Level, AudioStreamingState, VideoStreamingState, SystemContext of the app. 
+And regular hmi state is agnostic to system events. 
+A regular state can be changed by actiation\deactivation or resumption.
+
+There is a Temporary HMI state of the application.
+Temporary HMI state is not independent and should be on the top of Regular or another Temporary HMI state, consider it as parent state.
+Temporary HMI state does not contain any values, it contains the rules on how it changes values of the parent state. 
+The temporary state creates within SystemEvent start and deleted on SystemEvent end.
+
+SDL creates a linked list of HMI states for each app.
+This list contains at least one element: Regular HMIstate. And it always should be the first. 
+Temporary State added to this list on the system even start, and removed from this list on system event end. 
+
+System Events do not follow FILO, so events may be started on order 1, 2, 3, and ended on order 1, 3, 2.
```

Comment:
@LuxoftAKutsan I can share an image which I used in the past to explain how this "stack" works. Maybe you can re-use in this doc
![HmiStates](https://user-images.githubusercontent.com/5848997/81307659-e8743480-904e-11ea-81af-3dc5919bb8c0.png)

Reply:
Do you have sources of diagramm?

Reply:
@LuxoftAKutsan 
```
@startuml

title HMI states diagram for media app
[*] --> INVALID_ENUM
INVALID_ENUM: At SetInitialState()

note left
hmi_states: [
 0 - STATE_ID_REGULAR, INVALID_ENUM
]
current state: INVALID_ENUM, INVALID_ENUM
end note




INVALID_ENUM --> DEFAULT
DEFAULT: OnApplicationRegistered

note left
hmi_states: [
 0 - STATE_ID_REGULAR, NONE or BACKGROUND, NOT_AUDIBLE
]
current state: NONE or BACKGROUND, NOT_AUDIBLE
end note



DEFAULT --> FULL
FULL: OnAppActivated

note left
hmi_states: [
 0 - STATE_ID_REGULAR, FULL, AUDIBLE
]
current state: FULL, AUDIBLE
end note



FULL --> LIMITED
LIMITED: OnAppDeactivated

note right
hmi_states: [
 0 - STATE_ID_REGULAR, LIMITED, AUDIBLE
]
current state: LIMITED, AUDIBLE
end note



FULL --> BACKGROUND_1
BACKGROUND_1: HMIStateStarted<Phone call>

note right
hmi_states: [
 0 - STATE_ID_REGULAR, FULL, AUDIBLE
 1 - STATE_ID_PHONE_CALL (parent is item #0)
]
current state: BACKGROUND, NOT_AUDIBLE
end note



BACKGROUND_1 --> BACKGROUND_2
BACKGROUND_2: HMIStateStarted<VR Session>

note right
hmi_states: [
 0 - STATE_ID_REGULAR, FULL, AUDIBLE
 1 - STATE_ID_PHONE_CALL (parent is item #0)
 2 - STATE_ID_VR_SESSION (parent is item #1)
]
current state: BACKGROUND, NOT_AUDIBLE
end note



BACKGROUND_2 --> FULL_1
FULL_1: HMIStateStopped<Phone call>

note right
hmi_states: [
 0 - STATE_ID_REGULAR, FULL, AUDIBLE
 1 - STATE_ID_VR_SESSION (parent is item #0)
]
current state: FULL, NOT_AUDIBLE
end note



FULL_1 --> FULL_2
FULL_2: HMIStateStopped<VR session>

note left
hmi_states: [
 0 - STATE_ID_REGULAR, FULL, AUDIBLE
]
current state: FULL, NOT_AUDIBLE
end note



FULL_2 --> FULL
FULL_2 --> LIMITED
LIMITED --> FULL


@enduml
```

Reply:
I don't feel that this diagramm required in state controller SDD. I would say that it is partial and not clear without description. It is a good slide for presentation but not SDD diagram

Reply:
I think it would be nice to have diagram like this

Reply:
Diagram added :
https://www.plantuml.com/plantuml/img/nPNRRk8m48RlynGZhbfh-mAmgae9knjj21M7tXI4l6IMM78IopRLzkrtd51o08LjgsglIEpn_D_y7YPtnd9jzwc4Nf09AxY4eL9IRBaLUGPpoot2GlFC2BTWUfvdXPK8FoO16BThm1wt0V_mL4dql2odSrhi4Wjzw5K9GEyb7ObHOHORe8M3dB3qx9BQ5uMhREKj4c504SPsBxb-pGoq-3lNa68iE72bLXuXNR5hkBcvXQUllsxh-5kDH1YlWQ9SxhASlailyHabiJ4MH8_3CF0tusWvOMpWDla-5ld6M3gUYcotCUMvyy4pt1M4hq7r41YqNelAC5MvvlhbO60VdjP56LI2gu0I2j8kIw2gfja3cKKS3wXJOGmjZnAUxR1T5_55qsXCNCNJOZDS-i5zw1PEA5qWqTMYBiEL_cwjU2xOrarrnhujmv0n8F5A-An5Bf34ZyBGhLmAarIXD7STY_4LsPBa6OwubDTj9nYx7uv-VfzDbv5V4ZNj-0XKJTVnGMLDc_1nDiVYkluNtnSu8fIefwIvKXUzhEwrIVaXp1RPwIsNP5qUVag_rUwJHecFlEht4u0mc0IBiVzF38VOTs2yfV5TncPp3RoBi-2TdL_vHDC-GGtyy7iSpj1bNlS_uAo7rP7sw7ehShtDFiFI1cdZfNFKtT6Wfu_nNm00